### PR TITLE
Fix edgecase bugs in segmentation.relabel_sequential

### PR DESCRIPTION
**Pull request by uschmidt83**
## Description

- Fix edgecase bugs in `segmentation.relabel_sequential` (see modified tests).

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: skimage/segmentation/_join.py [hunk 1]
```diff
@@ -132,7 +132,9 @@ def relabel_sequential(label_field, offset=1):
     if np.dtype(required_type).itemsize > np.dtype(label_field.dtype).itemsize:
         label_field = label_field.astype(required_type)
     new_labels0 = np.arange(offset, offset + len(labels0))
-    if np.all(labels0 == new_labels0):
+    if offset == 1 and np.all(labels0 == new_labels0):
+        if not (labels == 0).any():
+            labels = np.concatenate(([0], labels))
         return label_field, labels, labels
```
Comment (uschmidt83) [inline skimage/segmentation/_join.py]: Maybe it would be best to not handle this special case separately and simply remove these lines of code.

File: skimage/segmentation/tests/test_join.py [hunk 2]
```diff
@@ -30,9 +30,15 @@ def test_join_segmentations():
         join_segmentations(s1, s3)
 
 
+def _check_maps(ar, ar_relab, fw, inv):
+    assert_array_equal(fw[ar], ar_relab)
+    assert_array_equal(inv[ar_relab], ar)
```
Comment (jni) [inline skimage/segmentation/tests/test_join.py]: Ooh I like this. =)

File: skimage/segmentation/_join.py (hunk 1 above)
Comment (jni) [reply skimage/segmentation/_join.py]: @uschmidt83 I'm actually confused about what's happening in this edge case. The test example has a lot going on so it's still not clear to me. Could you elaborate on the problem you're fixing and why it goes wrong in the current code?

Comment (uschmidt83) [reply skimage/segmentation/_join.py]: Hi @jni, these lines intended to check if `label_field` already contains sequential labels, hence there would be nothing to do. Unfortunately, the logic only works if `offset == 1` and `label_field` contains the 0 label. Hence, I added these things.

Comment (uschmidt83) [reply skimage/segmentation/_join.py]: This check was probably intended to avoid some computation, but it is unnecessary.
I just pushed a commit that (IMO) simplifies the function and makes it easier to understand and maintain.

File: skimage/segmentation/_join.py [hunk 3]
```diff
@@ -131,14 +131,11 @@ def relabel_sequential(label_field, offset=1):
     required_type = np.min_scalar_type(offset + len(labels0))
     if np.dtype(required_type).itemsize > np.dtype(label_field.dtype).itemsize:
         label_field = label_field.astype(required_type)
-    new_labels0 = np.arange(offset, offset + len(labels0))
-    if np.all(labels0 == new_labels0):
-        return label_field, labels, labels
+    new_m = offset - 1 + len(labels0)
```
Comment (jni) [inline skimage/segmentation/_join.py]: Is this short for `new_maximum` or `new_max_label`? If so, I wouldn't mind this being the new name rather than `new_m`.

Comment (uschmidt83) [reply skimage/segmentation/_join.py]: > Is this short for `new_maximum` or `new_max_label`?

Yes. After relabeling, `relabeled.max() == new_m`.

> If so, I wouldn't mind this being the new name rather than `new_m`.

I personally don't have an opinion on how to call this variable. I just chose the name because `m` is already used:

https://github.com/scikit-image/scikit-image/blob/7d70b5e7dd1f5e33326cb50b339aa0b5c13073c4/skimage/segmentation/_join.py#L124

Comment (jni) [reply skimage/segmentation/_join.py]: @uschmidt83 [who would write such code???](https://github.com/scikit-image/scikit-image/commit/dc377970f520f8f04a2b3d879983c3a1171518e7#diff-09660f97c6b4c3dc5d143d3be12e8aaaR85) :joy: :joy: :joy: 

Well, our preference has evolved to more self-explanatory variable names, so I would support a renaming of `m` to `max_label` and `new_m` to `new_max_label`. Feel free to leave that one to us, though, as indeed it is an orthogonal change to your contribution. Thank you for this fix!

Comment (uschmidt83) [reply skimage/segmentation/_join.py]: I just changed the variable names as requested.

File: skimage/segmentation/_join.py [hunk 4]
```diff
@@ -121,24 +121,21 @@ def relabel_sequential(label_field, offset=1):
...
+        new_type = np.min_scalar_type(int(max_label))
         label_field = label_field.astype(new_type)
-        m = m.astype(new_type)  # Ensures m is an integer
+        max_label = max_label.astype(new_type)  # Ensures max_label is an integer
     labels = np.unique(label_field)
     labels0 = labels[labels != 0]
-    required_type = np.min_scalar_type(offset + len(labels0))
+    new_max_label = offset - 1 + len(labels0)
+    new_labels0 = np.arange(offset, int(new_max_label + 1))
+    required_type = np.min_scalar_type(new_max_label)
```
Comment (uschmidt83) [inline skimage/segmentation/_join.py]: I also changed another subtlety regarding `required_type`.
(It is amazing how much complexity can be in such a small function.)

Comment (jni) [reply skimage/segmentation/_join.py]: >  (It is amazing how much complexity can be in such a small function.)

I know! Thank you again for your persistence here! It's a much better function thanks to your contributions!

File: skimage/segmentation/_join.py [hunk 5]
```diff
@@ -121,24 +121,21 @@ def relabel_sequential(label_field, offset=1):
...
-        new_type = np.min_scalar_type(int(m))
+        new_type = np.min_scalar_type(int(max_label))
         label_field = label_field.astype(new_type)
-        m = m.astype(new_type)  # Ensures m is an integer
+        max_label = max_label.astype(new_type)  # Ensures max_label is an integer
     labels = np.unique(label_field)
     labels0 = labels[labels != 0]
-    required_type = np.min_scalar_type(offset + len(labels0))
+    new_max_label = offset - 1 + len(labels0)
+    new_labels0 = np.arange(offset, int(new_max_label + 1))
```
Comment (rfezzani) [inline skimage/segmentation/_join.py]: No need for explicitely casting to int
```suggestion
    new_labels0 = np.arange(offset, new_max_label + 1)
```

File: skimage/segmentation/tests/test_join.py (hunk 2 above)
Comment (rfezzani) [reply skimage/segmentation/tests/test_join.py]: :+1:

File: skimage/segmentation/_join.py [hunk 6]
```diff
@@ -121,24 +121,21 @@ def relabel_sequential(label_field, offset=1):
...
+    required_type = np.min_scalar_type(new_max_label)
     if np.dtype(required_type).itemsize > np.dtype(label_field.dtype).itemsize:
         label_field = label_field.astype(required_type)
-    new_labels0 = np.arange(offset, offset + len(labels0))
-    if np.all(labels0 == new_labels0):
-        return label_field, labels, labels
-    forward_map = np.zeros(int(m + 1), dtype=label_field.dtype)
+    forward_map = np.zeros(int(max_label + 1), dtype=label_field.dtype)
```
Comment (rfezzani) [inline skimage/segmentation/_join.py]: I think that you can directly use `new_type` here and avoid the `if` above
```suggestion
    forward_map = np.zeros(int(max_label + 1), dtype=new_type)
```

File: skimage/segmentation/_join.py [hunk 7]
```diff
@@ -121,24 +121,21 @@ def relabel_sequential(label_field, offset=1):
...
     forward_map[labels0] = new_labels0
-    if not (labels == 0).any():
-        labels = np.concatenate(([0], labels))
-    inverse_map = np.zeros(offset - 1 + len(labels), dtype=label_field.dtype)
-    inverse_map[(offset - 1):] = labels
+    inverse_map = np.zeros(int(new_max_label + 1), dtype=label_field.dtype)
```
Comment (rfezzani) [inline skimage/segmentation/_join.py]: Here again
```suggestion
    inverse_map = np.zeros(int(new_max_label + 1), dtype=new_type)
```

File: skimage/segmentation/_join.py (hunk 5 above)
Comment (uschmidt83) [reply skimage/segmentation/_join.py]: Hi @rfezzani, you are right about that. I was wrongly assuming that `int(v)` would not change the data type of `v` if it was "integer-based" (e.g. `np.uint16`).

I went ahead and changed the code some more, as seen in the latest commit.

File: skimage/segmentation/_join.py (hunk 6 above)
Comment (uschmidt83) [reply skimage/segmentation/_join.py]: > I think that you can directly use `required_type` here and avoid the `if` above

In my last pull request (#3740), I changed the function to retain the data type of the input `label_field`. Hence, I would keep that. But you are right that it is not necessary to convert (i.e. make a copy of) `label_field` to the new data type. I changed that in the latest commit.
